docs: Replace Middleware docs to Proxy

devjiwonchoi opened the pull request:
> [!NOTE]
> Best reviewed by each commit for better diff view.

This PR clones the Middleware docs for Proxy and removes the Middleware docs.

Did not clone the list of docs:

- `errors/middleware-upgrade.mdx` - It's a middleware upgrade guide from v12.2
- `errors/beta-middleware.mdx` - It's an error when using middleware before v12.2
- `errors/returning-response-body.mdx` - Legacy behavior from versions < v12.2

devjiwonchoi committed:
Clean-clone Middleware docs for Proxy

devjiwonchoi committed:
Replace terms Middleware to Proxy

devjiwonchoi commented:
* **#84964** <a href="https://app.graphite.dev/github/pr/vercel/next.js/84964?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#84711** <a href="https://app.graphite.dev/github/pr/vercel/next.js/84711?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#84710** <a href="https://app.graphite.dev/github/pr/vercel/next.js/84710?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#84709** <a href="https://app.graphite.dev/github/pr/vercel/next.js/84709?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/84709?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#84708** <a href="https://app.graphite.dev/github/pr/vercel/next.js/84708?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `canary`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

icyJoseph commented on errors/deleting-query-params-in-proxy.mdx:
```diff
@@ -0,0 +1,37 @@
+
+export default function proxy(request: NextRequest) {
+  const nextUrl = request.nextUrl
+  nextUrl.searchParams.delete('key') // <-- this is now possible! 🎉
```
```suggestion
  nextUrl.searchParams.delete('key') // <-- this is now possible!
```

No emojis ~

devjiwonchoi replied:
It's a clean copy from middleware docs 😂 Remove from both?

https://github.com/vercel/next.js/blob/c6cd1f455f5c55530ce894aad3210ef5f6003a77/errors/deleting-query-params-in-middlewares.mdx

icyJoseph replied:
🙏 yes haha

devjiwonchoi committed:
Address comments suggestion for fixing in both proxy and middleware docs

Co-authored-by: Joseph <sephxd1234@gmail.com>

icyJoseph commented on docs/01-app/03-api-reference/03-file-conventions/proxy.mdx:
```diff
@@ -0,0 +1,697 @@
+
+## Version history
+
+| Version   | Changes                                                                                    |
```
I guess we can keep the Middleware in the history here, and add a v16 row, with the renaming

If one uses v14, or v13, it'll be middleware

devjiwonchoi committed:
Address comments
- Example of "Set-Cookie:vercel=fast;path=/"
- keep the Middleware in the history

devjiwonchoi committed:
Address comment https://github.com/vercel/next.js/pull/84709#discussion_r2439326565

icyJoseph commented on docs/01-app/03-api-reference/03-file-conventions/proxy.mdx:
```diff
@@ -0,0 +1,698 @@
+    - app/api-reference/functions/next-response
+---
+
+The `proxy.js|ts` file is used to write [Proxy](/docs/app/getting-started/route-handlers-and-proxy#proxy) and run code on the server before a request is completed. Then, based on the incoming request, you can modify the response by rewriting, redirecting, modifying the request or response headers, or responding directly.
```
I guess this link is gonna be broken, since proxy is separate from route handlers in: https://github.com/vercel/next.js/pull/84708/files

devjiwonchoi replied:
Oh good catch, will replace

devjiwonchoi committed:
Update route-handlers-and-middleware link

Co-authored-by: Joseph <sephxd1234@gmail.com>

devjiwonchoi commented:
Noticed the docs link validation was accurate 😅 should we replace the doc below at https://github.com/vercel/next.js/pull/84964?

https://github.com/vercel/next.js/blob/8bc8dc343c30fdc263fcf719fbc99febe0fe4c93/docs/01-app/02-guides/self-hosting.mdx#L25-L29

EDIT: Added to https://github.com/vercel/next.js/pull/84964
